internal/alb/lb/loadbalancer.go

```diff
@@ -321,3 +334,143 @@ func (controller *defaultController) validateLBConfig(ctx context.Context, ingre
 
 	return nil
 }
+
+func (controller *defaultController) resolveSecurityGroupNames(ctx context.Context, sgIDOrNames []string) ([]string, error) {
+	var names []string
+	var output []string
+
+	for _, sg := range sgIDOrNames {
+		if strings.HasPrefix(sg, "sg-") {
+			output = append(output, sg)
+			continue
+		}
+
+		names = append(names, sg)
+	}
+
+	if len(names) > 0 {
+		groups, err := controller.cloud.GetSecurityGroupsByName(ctx, names)
+		if err != nil {
+			return output, err
+		}
+
+		for _, sg := range groups {
+			output = append(output, aws.StringValue(sg.GroupId))
+		}
+	}
+
+	if len(output) != len(sgIDOrNames) {
+		return output, fmt.Errorf("not all security groups were resolvable, (%v != %v)", strings.Join(sgIDOrNames, ","), strings.Join(output, ","))
+	}
+
+	return output, nil
+}
+
+func (controller *defaultController) resolveSubnets(ctx context.Context, scheme string, in []string) ([]string, error) {
+	if len(in) == 0 {
+		subnets, err := controller.clusterSubnets(ctx, scheme)
+		return subnets, err
+
+	}
+
+	var names []string
+	var subnets []string
+
+	for _, subnet := range in {
+		if strings.HasPrefix(subnet, "subnet-") {
+			subnets = append(subnets, subnet)
+			continue
+		}
+		names = append(names, subnet)
+	}
+
+	if len(names) > 0 {
+		o, err := controller.cloud.GetSubnetsByNameOrID(ctx, names)
+		if err != nil {
+			return subnets, err
+		}
+
+		for _, subnet := range o {
+			subnets = append(subnets, aws.StringValue(subnet.SubnetId))
+		}
+	}
+
+	sort.Strings(subnets)
+	if len(subnets) != len(in) {
+		return subnets, fmt.Errorf("not all subnets were resolvable, (%v != %v)", strings.Join(in, ","), strings.Join(subnets, ","))
+	}
+
+	return subnets, nil
+}
+
+func (controller *defaultController) clusterSubnets(ctx context.Context, scheme string) ([]string, error) {
+	var useableSubnets []*ec2.Subnet
+	var out []string
+	var key string
+
+	if scheme == elbv2.LoadBalancerSchemeEnumInternal {
+		key = aws.TagNameSubnetInternalELB
+	} else if scheme == elbv2.LoadBalancerSchemeEnumInternetFacing {
+		key = aws.TagNameSubnetPublicELB
+	} else {
+		return nil, fmt.Errorf("invalid scheme [%s]", scheme)
+	}
+
+	clusterSubnets, err := controller.cloud.GetClusterSubnets()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get AWS tags. Error: %s", err.Error())
+	}
+
+	var subnetIds []string
+	for arn, subnetTags := range clusterSubnets {
+		for _, tag := range subnetTags {
+			if aws.StringValue(tag.Key) == key {
+				p := strings.Split(arn, "/")
+				subnetID := p[len(p)-1]
+				subnetIds = append(subnetIds, subnetID)
+			}
+		}
+	}
+
+	if len(subnetIds) == 0 {
+		sort.Strings(out)
```

Comment:
seems unneeded 😄

Reply:
the sorting helps when doing comparisons for changes

Reply:
but seems out is empty at this branch 😅

Reply:
now i see :)